habitat_code/kobylarz_kmeans_review_code_20240313.py

```diff
@@ -0,0 +1,603 @@
+                   'band_number': [band]
+                  })
+        granule_df = pd.concat([granule_df, new_row], ignore_index=True)
+    return granule_df
+
+
+# In[8]:
+
+
+@cached("download_df", False)
+def do_something(results, open_results):
+    download_df = create_granule_df(results, open_results)
+    return download_df
+download_df = do_something(results, open_results)
+
+
+# ### Open, crop, and mask data
+# 
+# This will be the most resource-intensive step. I recommend caching your results using the `cached` decorator or by writing your own caching code. I also recommend testing this step with one or two dates before running the full computation.
+# 
+# This code should include at least one **function** including a numpy-style docstring. A good place to start would be a function for opening a single masked raster, applying the appropriate scale parameter, and cropping.
+# 
+# **YOUR TASK:**
+# 1. For each granule:
+#     1. Open the Fmask band, crop, and compute a quality mask for the granule. You can use the following code as a starting point, making sure that `mask_bits` contains the quality bits you want to consider:
+#          ```python
+#          # Expand into a new dimension of binary bits
+#          bits = np.unpackbits(da.astype(np.uint8)).reshape(da.shape + (-1,))
+# 
+#          # Select the required bits and check if any are flagged
+#          mask = np.prod(bits[..., mask_bits]==0, axis=-1)
+#          ```
+# 
+#     2. For each band that starts with 'B':
+#         1. Open the band, crop, and apply the scale factor
+#         2. Name the DataArray after the band using the `.name` attribute
+#         3. Apply the cloud mask using the `.where()` method
+#         4. Store the DataArray in your data structure (e.g. adding a GeoDataFrame column with the DataArray in it. Note that you will need to remove the rows for unused bands)
+
+# In[9]:
+
+
+# Function to compute image mask
+def process_tifs(download_df, huc_gdf):
+    """
+    Load and process the Fmask and B band .tif files
+    in a granule.
+
+    Parameters
+    ==========
+    download_df : pandas.core.groupby.generic.DataFrame
+      Dataframe containing all the rows for each band .tif for all the files.
+    
+    huc_gdf : GeoDataFrame
+      Geodataframe containing the geometry of the HUC watershed used to clip
+
+    Returns
+    =======
+    accumulator_df : Pandas dataframe
+      Dataframe containing the processed rows for all the tifs.
+      An xarray has been added for each processed .tif in the last column.
+    """
+    grouped = download_df.groupby('granule_id')
+    columns = ['geometry',
+               'granule_id', 
+               'file_url', 
+               'tile_id', 
+               'datetime', 
+               'band_number',
+               'processed_da']
+    accumulator_df = pd.DataFrame(columns=columns)
+    # Looping through each Granule and processing data
+    # Create cloud mask
+    for granule_id, group_data in grouped:
```

Comment:
I think this section needs more comments, or at least to be broken up a bit. There's a lot going on here, and I think you'll find it easier to read if you come back to it if there's a little more guidance about what's happening.